File: db4on/core/src/core/com/db4o/YapClassCollection.cs
namespace com.db4o
{
	/// <exclude></exclude>
	public sealed class YapClassCollection : com.db4o.YapMeta
	{
		private com.db4o.foundation.Collection4 i_classes;

		private com.db4o.foundation.Hashtable4 i_creating;

		private readonly com.db4o.Transaction _systemTransaction;

		private com.db4o.foundation.Hashtable4 i_yapClassByBytes;

		private com.db4o.foundation.Hashtable4 i_yapClassByClass;

		private com.db4o.foundation.Hashtable4 i_yapClassByID;

		private int i_yapClassCreationDepth;

		private com.db4o.foundation.Queue4 i_initYapClassesOnUp;

		private readonly com.db4o.PendingClassInits _classInits;

		internal YapClassCollection(com.db4o.Transaction systemTransaction)
		{
			_systemTransaction = systemTransaction;
			i_initYapClassesOnUp = new com.db4o.foundation.Queue4();
			_classInits = new com.db4o.PendingClassInits(_systemTransaction);
		}

		public void AddYapClass(com.db4o.YapClass yapClass)
		{
			Stream().SetDirtyInSystemTransaction(this);
			i_classes.Add(yapClass);
			if (yapClass.StateUnread())
			{
				i_yapClassByBytes.Put(yapClass.i_nameBytes, yapClass);
			}
			else
			{
				i_yapClassByClass.Put(yapClass.ClassReflector(), yapClass);
			}
			if (yapClass.GetID() == 0)
			{
				yapClass.Write(_systemTransaction);
			}
			i_yapClassByID.Put(yapClass.GetID(), yapClass);
		}

		private byte[] AsBytes(string str)
		{
			return Stream().StringIO().Write(str);
		}

		internal void AttachQueryNode(string fieldName, com.db4o.foundation.Visitor4 a_visitor
			)
		{
			com.db4o.YapClassCollectionIterator i = Iterator();
			while (i.MoveNext())
			{
				com.db4o.YapClass yc = i.CurrentClass();
				if (!yc.IsInternal())
				{
					yc.ForEachYapField(new _AnonymousInnerClass59(this, fieldName, a_visitor, yc));
				}
			}
		}

		private sealed class _AnonymousInnerClass59 : com.db4o.foundation.Visitor4
		{
			public _AnonymousInnerClass59(YapClassCollection _enclosing, string fieldName, com.db4o.foundation.Visitor4
				 a_visitor, com.db4o.YapClass yc)
			{
				this._enclosing = _enclosing;
				this.fieldName = fieldName;
				this.a_visitor = a_visitor;
				this.yc = yc;
			}

			public void Visit(object obj)
			{
				com.db4o.YapField yf = (com.db4o.YapField)obj;
				if (yf.CanAddToQuery(fieldName))
				{
					a_visitor.Visit(new object[] { yc, yf });
				}
			}

			private readonly YapClassCollection _enclosing;

			private readonly string fieldName;

			private readonly com.db4o.foundation.Visitor4 a_visitor;

			private readonly com.db4o.YapClass yc;
		}

		internal void CheckChanges()
		{
			System.Collections.IEnumerator i = i_classes.GetEnumerator();
			while (i.MoveNext())
			{
				((com.db4o.YapClass)i.Current).CheckChanges();
			}
		}

		internal bool CreateYapClass(com.db4o.YapClass a_yapClass, com.db4o.reflect.ReflectClass
			 a_class)
		{
			i_yapClassCreationDepth++;
			com.db4o.reflect.ReflectClass superClass = a_class.GetSuperclass();
			com.db4o.YapClass superYapClass = null;
			if (superClass != null && !superClass.Equals(Stream().i_handlers.ICLASS_OBJECT))
			{
				superYapClass = GetYapClass(superClass, true);
			}
			bool ret = Stream().CreateYapClass(a_yapClass, a_class, superYapClass);
			i_yapClassCreationDepth--;
			InitYapClassesOnUp();
			return ret;
		}

		internal bool FieldExists(string a_field)
		{
			com.db4o.YapClassCollectionIterator i = Iterator();
			while (i.MoveNext())
			{
				if (i.CurrentClass().GetYapField(a_field) != null)
				{
					return true;
				}
			}
			return false;
		}

		internal com.db4o.foundation.Collection4 ForInterface(com.db4o.reflect.ReflectClass
			 claxx)
		{
			com.db4o.foundation.Collection4 col = new com.db4o.foundation.Collection4();
			com.db4o.YapClassCollectionIterator i = Iterator();
			while (i.MoveNext())
			{
				com.db4o.YapClass yc = i.CurrentClass();
				com.db4o.reflect.ReflectClass candidate = yc.ClassReflector();
				if (!candidate.IsInterface())
				{
					if (claxx.IsAssignableFrom(candidate))
					{
						col.Add(yc);
						System.Collections.IEnumerator j = new com.db4o.foundation.Collection4(col).GetEnumerator
							();
						while (j.MoveNext())
						{
							com.db4o.YapClass existing = (com.db4o.YapClass)j.Current;
							if (existing != yc)
							{
								com.db4o.YapClass higher = yc.GetHigherHierarchy(existing);
								if (higher != null)
								{
									if (higher == yc)
									{
										col.Remove(existing);
									}
									else
									{
										col.Remove(yc);
									}
								}
							}
						}
					}
				}
			}
			return col;
		}

		public override byte GetIdentifier()
		{
			return com.db4o.YapConst.YAPCLASSCOLLECTION;
		}

		internal com.db4o.YapClass GetActiveYapClass(com.db4o.reflect.ReflectClass a_class
			)
		{
			return (com.db4o.YapClass)i_yapClassByClass.Get(a_class);
		}

		internal com.db4o.YapClass GetYapClass(com.db4o.reflect.ReflectClass a_class, bool
			 a_create)
		{
			com.db4o.YapClass yapClass = (com.db4o.YapClass)i_yapClassByClass.Get(a_class);
			if (yapClass == null)
			{
				yapClass = (com.db4o.YapClass)i_yapClassByBytes.Remove(GetNameBytes(a_class.GetName
					()));
				ReadYapClass(yapClass, a_class);
			}
			if (yapClass != null || (!a_create))
			{
				return yapClass;
			}
			yapClass = (com.db4o.YapClass)i_creating.Get(a_class);
			if (yapClass != null)
			{
				return yapClass;
			}
			yapClass = new com.db4o.YapClass(Stream(), a_class);
			i_creating.Put(a_class, yapClass);
			if (!CreateYapClass(yapClass, a_class))
			{
				i_creating.Remove(a_class);
				return null;
			}
			bool addMembers = false;
			if (i_yapClassByClass.Get(a_class) == null)
			{
				AddYapClass(yapClass);
				addMembers = true;
			}
			int id = yapClass.GetID();
			if (id == 0)
			{
				yapClass.Write(Stream().GetSystemTransaction());
				id = yapClass.GetID();
			}
			if (i_yapClassByID.Get(id) == null)
			{
				i_yapClassByID.Put(id, yapClass);
				addMembers = true;
			}
			if (addMembers || yapClass.i_fields == null)
			{
				_classInits.Process(yapClass);
			}
			i_creating.Remove(a_class);
			Stream().SetDirtyInSystemTransaction(this);
			return yapClass;
		}

		internal com.db4o.YapClass GetYapClass(int a_id)
		{
			return ReadYapClass((com.db4o.YapClass)i_yapClassByID.Get(a_id), null);
		}

		public com.db4o.YapClass GetYapClass(string a_name)
		{
			com.db4o.YapClass yapClass = (com.db4o.YapClass)i_yapClassByBytes.Remove(GetNameBytes
				(a_name));
			ReadYapClass(yapClass, null);
			if (yapClass == null)
			{
				com.db4o.YapClassCollectionIterator i = Iterator();
				while (i.MoveNext())
				{
					yapClass = (com.db4o.YapClass)i.Current;
					if (a_name.Equals(yapClass.GetName()))
					{
						ReadYapClass(yapClass, null);
						return yapClass;
					}
				}
				return null;
			}
			return yapClass;
		}

		public int GetYapClassID(string name)
		{
			com.db4o.YapClass yc = (com.db4o.YapClass)i_yapClassByBytes.Get(GetNameBytes(name
				));
			if (yc != null)
			{
				return yc.GetID();
			}
			return 0;
		}

		private byte[] GetNameBytes(string name)
		{
			return AsBytes(ResolveAlias(name));
		}

		private string ResolveAlias(string name)
		{
			return Stream().ConfigImpl().ResolveAlias(name);
		}

		internal void InitOnUp(com.db4o.Transaction systemTrans)
		{
			i_yapClassCreationDepth++;
			systemTrans.Stream().ShowInternalClasses(true);
			System.Collections.IEnumerator i = i_classes.GetEnumerator();
			while (i.MoveNext())
			{
				((com.db4o.YapClass)i.Current).InitOnUp(systemTrans);
			}
			systemTrans.Stream().ShowInternalClasses(false);
			i_yapClassCreationDepth--;
			InitYapClassesOnUp();
		}

		internal void InitTables(int a_size)
		{
			i_classes = new com.db4o.foundation.Collection4();
			i_yapClassByBytes = new com.db4o.foundation.Hashtable4(a_size);
			if (a_size < 16)
			{
				a_size = 16;
			}
			i_yapClassByClass = new com.db4o.foundation.Hashtable4(a_size);
			i_yapClassByID = new com.db4o.foundation.Hashtable4(a_size);
			i_creating = new com.db4o.foundation.Hashtable4(1);
		}

		private void InitYapClassesOnUp()
		{
			if (i_yapClassCreationDepth == 0)
			{
				com.db4o.YapClass yc = (com.db4o.YapClass)i_initYapClassesOnUp.Next();
				while (yc != null)
				{
					yc.InitOnUp(_systemTransaction);
					yc = (com.db4o.YapClass)i_initYapClassesOnUp.Next();
				}
			}
		}

		public com.db4o.YapClassCollectionIterator Iterator()
		{
			return new com.db4o.YapClassCollectionIterator(this, new com.db4o.foundation.ArrayIterator4
				(i_classes.ToArray()));
		}

		private class ClassIDIterator : com.db4o.foundation.MappingIterator
		{
			public ClassIDIterator(com.db4o.foundation.Collection4 classes) : base(classes.GetEnumerator
				())
			{
			}

			protected override object Map(object current)
			{
				return ((com.db4o.YapClass)current).GetID();
			}
		}

		public System.Collections.IEnumerator Ids()
		{
			return new com.db4o.YapClassCollection.ClassIDIterator(i_classes);
		}

		public override int OwnLength()
		{
			return com.db4o.YapConst.OBJECT_LENGTH + com.db4o.YapConst.INT_LENGTH + (i_classes
				.Size() * com.db4o.YapConst.ID_LENGTH);
		}

		internal void Purge()
		{
			System.Collections.IEnumerator i = i_classes.GetEnumerator();
			while (i.MoveNext())
			{
				((com.db4o.YapClass)i.Current).Purge();
			}
		}

		public sealed override void ReadThis(com.db4o.Transaction a_trans, com.db4o.YapReader
			 a_reader)
		{
			int classCount = a_reader.ReadInt();
			InitTables(classCount);
			for (int i = classCount; i > 0; i--)
			{
				com.db4o.YapClass yapClass = new com.db4o.YapClass(Stream(), null);
				int id = a_reader.ReadInt();
				yapClass.SetID(id);
				i_classes.Add(yapClass);
				i_yapClassByID.Put(id, yapClass);
				i_yapClassByBytes.Put(yapClass.ReadName(a_trans), yapClass);
			}
			ApplyReadAs();
		}

		private void ApplyReadAs()
		{
			com.db4o.foundation.Hashtable4 readAs = Stream().ConfigImpl().ReadAs();
			readAs.ForEachKey(new _AnonymousInnerClass321(this, readAs));
		}

		private sealed class _AnonymousInnerClass321 : com.db4o.foundation.Visitor4
		{
			public _AnonymousInnerClass321(YapClassCollection _enclosing, com.db4o.foundation.Hashtable4
				 readAs)
			{
				this._enclosing = _enclosing;
				this.readAs = readAs;
			}

			public void Visit(object a_object)
			{
				string dbName = (string)a_object;
				byte[] dbbytes = this._enclosing.GetNameBytes(dbName);
				string useName = (string)readAs.Get(dbName);
				byte[] useBytes = this._enclosing.GetNameBytes(useName);
				if (this._enclosing.i_yapClassByBytes.Get(useBytes) == null)
				{
					com.db4o.YapClass yc = (com.db4o.YapClass)this._enclosing.i_yapClassByBytes.Get(dbbytes
						);
					if (yc != null)
					{
						yc.i_nameBytes = useBytes;
						yc.SetConfig(this._enclosing.Stream().ConfigImpl().ConfigClass(dbName));
						this._enclosing.i_yapClassByBytes.Put(dbbytes, null);
						this._enclosing.i_yapClassByBytes.Put(useBytes, yc);
					}
				}
			}

			private readonly YapClassCollection _enclosing;

			private readonly com.db4o.foundation.Hashtable4 readAs;
		}

		public com.db4o.YapClass ReadYapClass(com.db4o.YapClass yapClass, com.db4o.reflect.ReflectClass
			 a_class)
		{
			if (yapClass != null && !yapClass.StateUnread())
			{
				return yapClass;
			}
			i_yapClassCreationDepth++;
			if (yapClass != null && yapClass.StateUnread())
			{
				yapClass.CreateConfigAndConstructor(i_yapClassByBytes, Stream(), a_class);
				com.db4o.reflect.ReflectClass claxx = yapClass.ClassReflector();
				if (claxx != null)
				{
					i_yapClassByClass.Put(claxx, yapClass);
					yapClass.ReadThis();
					yapClass.CheckChanges();
					i_initYapClassesOnUp.Add(yapClass);
				}
			}
			i_yapClassCreationDepth--;
			InitYapClassesOnUp();
			return yapClass;
		}

		public void RefreshClasses()
		{
			com.db4o.YapClassCollection rereader = new com.db4o.YapClassCollection(_systemTransaction
				);
			rereader.i_id = i_id;
			rereader.Read(Stream().GetSystemTransaction());
			System.Collections.IEnumerator i = rereader.i_classes.GetEnumerator();
			while (i.MoveNext())
			{
				com.db4o.YapClass yc = (com.db4o.YapClass)i.Current;
				if (i_yapClassByID.Get(yc.GetID()) == null)
				{
					i_classes.Add(yc);
					i_yapClassByID.Put(yc.GetID(), yc);
					if (yc.StateUnread())
					{
						i_yapClassByBytes.Put(yc.ReadName(_systemTransaction), yc);
					}
					else
					{
						i_yapClassByClass.Put(yc.ClassReflector(), yc);
					}
				}
			}
			i = i_classes.GetEnumerator();
			while (i.MoveNext())
			{
				com.db4o.YapClass yc = (com.db4o.YapClass)i.Current;
				yc.Refresh();
			}
		}

		internal void ReReadYapClass(com.db4o.YapClass yapClass)
		{
			if (yapClass != null)
			{
				ReReadYapClass(yapClass.i_ancestor);
				yapClass.ReadName(_systemTransaction);
				yapClass.ForceRead();
				yapClass.SetStateClean();
				yapClass.BitFalse(com.db4o.YapConst.CHECKED_CHANGES);
				yapClass.BitFalse(com.db4o.YapConst.READING);
				yapClass.BitFalse(com.db4o.YapConst.CONTINUE);
				yapClass.BitFalse(com.db4o.YapConst.DEAD);
				yapClass.CheckChanges();
			}
		}

		public com.db4o.ext.StoredClass[] StoredClasses()
		{
			com.db4o.foundation.Collection4 classes = new com.db4o.foundation.Collection4();
			System.Collections.IEnumerator i = i_classes.GetEnumerator();
			while (i.MoveNext())
			{
				com.db4o.YapClass yc = (com.db4o.YapClass)i.Current;
				ReadYapClass(yc, null);
				if (yc.ClassReflector() == null)
				{
					yc.ForceRead();
				}
				classes.Add(yc);
			}
			ApplyReadAs();
			com.db4o.ext.StoredClass[] sclasses = new com.db4o.ext.StoredClass[classes.Size()
				];
			classes.ToArray(sclasses);
			return sclasses;
		}

		public void WriteAllClasses()
		{
			com.db4o.ext.StoredClass[] storedClasses = StoredClasses();
			for (int i = 0; i < storedClasses.Length; i++)
			{
				com.db4o.YapClass yc = (com.db4o.YapClass)storedClasses[i];
				yc.SetStateDirty();
			}
			for (int i = 0; i < storedClasses.Length; i++)
			{
				com.db4o.YapClass yc = (com.db4o.YapClass)storedClasses[i];
				yc.Write(_systemTransaction);
			}
		}

		public override void WriteThis(com.db4o.Transaction trans, com.db4o.YapReader a_writer
			)
		{
			a_writer.WriteInt(i_classes.Size());
			System.Collections.IEnumerator i = i_classes.GetEnumerator();
			while (i.MoveNext())
			{
				a_writer.WriteIDOf(trans, i.Current);
			}
		}

		public override string ToString()
		{
			return base.ToString();
			string str = string.Empty;
			System.Collections.IEnumerator i = i_classes.GetEnumerator();
			while (i.MoveNext())
			{
				com.db4o.YapClass yc = (com.db4o.YapClass)i.Current;
				str += yc.GetID() + " " + yc + "\r\n";
			}
			return str;
		}

		public static void Defrag(com.db4o.ReaderPair readers)
		{
			int numClasses = readers.ReadInt();
			for (int classIdx = 0; classIdx < numClasses; classIdx++)
			{
				readers.CopyID();
			}
		}

		private com.db4o.YapStream Stream()
		{
			return _systemTransaction.Stream();
		}

		public override void SetID(int a_id)
		{
			if (Stream().IsClient())
			{
				base.SetID(a_id);
				return;
			}
			if (i_id == 0)
			{
				SystemData().ClassCollectionID(a_id);
			}
			base.SetID(a_id);
		}

		private com.db4o.inside.SystemData SystemData()
		{
			return _systemTransaction.i_file.SystemData();
		}
	}
}

File: db4oj/core/src/com/db4o/YapClassCollection.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o;

import com.db4o.ext.StoredClass;
import com.db4o.foundation.ArrayIterator4;
import com.db4o.foundation.Collection4;
import com.db4o.foundation.Debug4;
import com.db4o.foundation.Hashtable4;
import com.db4o.foundation.Iterator4;
import com.db4o.foundation.MappingIterator;
import com.db4o.foundation.Queue4;
import com.db4o.foundation.Visitor4;
import com.db4o.inside.*;
import com.db4o.reflect.ReflectClass;

/**
 * @exclude
 */
public final class YapClassCollection extends YapMeta {

    private Collection4 i_classes;
    private Hashtable4 i_creating;
    
    private final Transaction _systemTransaction;

    private Hashtable4 i_yapClassByBytes;
    private Hashtable4 i_yapClassByClass;
    private Hashtable4 i_yapClassByID;
    
    private int i_yapClassCreationDepth;
    private Queue4 i_initYapClassesOnUp;
	
	private final PendingClassInits _classInits; 


    YapClassCollection(Transaction systemTransaction) {
        _systemTransaction = systemTransaction;
        i_initYapClassesOnUp = new Queue4();
		_classInits = new PendingClassInits(_systemTransaction);
    }

    public void addYapClass(YapClass yapClass) {
        stream().setDirtyInSystemTransaction(this);
        i_classes.add(yapClass);
        if(yapClass.stateUnread()){
            i_yapClassByBytes.put(yapClass.i_nameBytes, yapClass);
        }else{
            i_yapClassByClass.put(yapClass.classReflector(), yapClass);
        }
        if (yapClass.getID() == 0) {
            yapClass.write(_systemTransaction);
        }
        i_yapClassByID.put(yapClass.getID(), yapClass);
    }
    
    private byte[] asBytes(String str){
        return stream().stringIO().write(str);
    }

    void attachQueryNode(final String fieldName, final Visitor4 a_visitor) {
        YapClassCollectionIterator i = iterator();
        while (i.moveNext()) {
            final YapClass yc = i.currentClass();
            if(! yc.isInternal()){
                yc.forEachYapField(new Visitor4() {
                    public void visit(Object obj) {
                        YapField yf = (YapField)obj;
                        if(yf.canAddToQuery(fieldName)){
                            a_visitor.visit(new Object[] {yc, yf});
                        }
                    }
                });
            }
        }
    }

    void checkChanges() {
        Iterator4 i = i_classes.iterator();
        while (i.moveNext()) {
            ((YapClass)i.current()).checkChanges();
        }
    }
    
    final boolean createYapClass(YapClass a_yapClass, ReflectClass a_class) {
        i_yapClassCreationDepth++;
        ReflectClass superClass = a_class.getSuperclass();
        YapClass superYapClass = null;
        if (superClass != null && ! superClass.equals(stream().i_handlers.ICLASS_OBJECT)) {
            superYapClass = produceYapClass(superClass);
        }
        boolean ret = stream().createYapClass(a_yapClass, a_class, superYapClass);
        i_yapClassCreationDepth--;
        initYapClassesOnUp();
        return ret;
    }

	public static void defrag(ReaderPair readers) {
        if (Deploy.debug) {
            readers.readBegin(YapConst.YAPCLASSCOLLECTION);
        }
		int numClasses=readers.readInt();
		for(int classIdx=0;classIdx<numClasses;classIdx++) {
			readers.copyID();
		}
        if (Deploy.debug) {
            readers.readEnd();
        }
	}

	private void ensureAllClassesRead() {
		boolean allClassesRead=false;
    	while(!allClassesRead) {
	    	Collection4 unreadClasses=new Collection4();
			int numClasses=i_classes.size();
	        Iterator4 classIter = i_classes.iterator();
	        while(classIter.moveNext()) {
	        	YapClass yapClass=(YapClass)classIter.current();
	        	if(yapClass.stateUnread()) {
	        		unreadClasses.add(yapClass);
	        	}
	        }
	        Iterator4 unreadIter=unreadClasses.iterator();
	        while(unreadIter.moveNext()) {
	        	YapClass yapClass=(YapClass)unreadIter.current();
	        	readYapClass(yapClass,null);
	            if(yapClass.classReflector() == null){
	            	yapClass.forceRead();
	            }
	        }
	        allClassesRead=(i_classes.size()==numClasses);
    	}
		applyReadAs();
	}

    boolean fieldExists(String a_field) {
        YapClassCollectionIterator i = iterator();
        while (i.moveNext()) {
            if (i.currentClass().getYapField(a_field) != null) {
                return true;
            }
        }
        return false;
    }

    Collection4 forInterface(ReflectClass claxx) {
        Collection4 col = new Collection4();
        YapClassCollectionIterator i = iterator();
        while (i.moveNext()) {
            YapClass yc = i.currentClass();
            ReflectClass candidate = yc.classReflector();
            if(! candidate.isInterface()){
                if (claxx.isAssignableFrom(candidate)) {
                    col.add(yc);
                    Iterator4 j = new Collection4(col).iterator();
                    while (j.moveNext()) {
                        YapClass existing = (YapClass)j.current();
                        if(existing != yc){
                            YapClass higher = yc.getHigherHierarchy(existing);
                            if (higher != null) {
                                if (higher == yc) {
                                    col.remove(existing);
                                }else{
                                    col.remove(yc);
                                }
                            }
                        }
                    }
                }
            }
        }
        return col;
    }

    public byte getIdentifier() {
        return YapConst.YAPCLASSCOLLECTION;
    }
    
    YapClass getActiveYapClass(ReflectClass a_class) {
        return (YapClass)i_yapClassByClass.get(a_class);
    }
    
    YapClass getYapClass (ReflectClass a_class) {
    	YapClass yapClass = (YapClass)i_yapClassByClass.get(a_class);
        if (yapClass != null) {
        	return yapClass;
        }
        yapClass = (YapClass)i_yapClassByBytes.remove(getNameBytes(a_class.getName()));
        readYapClass(yapClass, a_class);
        return yapClass;
    }

    YapClass produceYapClass(ReflectClass a_class) {
    	
    	YapClass yapClass = getYapClass(a_class);
    	
        if (yapClass != null ) {
            return yapClass;
        }
        
        yapClass = (YapClass)i_creating.get(a_class);
        
        if(yapClass != null){
            return yapClass;
        }
        
        yapClass = new YapClass(stream(), a_class);
        
        i_creating.put(a_class, yapClass);
        
        if(! createYapClass(yapClass, a_class)){
            i_creating.remove(a_class);
            return null;
        }

        // YapStream#createYapClass may add the YapClass already,
        // so we have to check again
        
        boolean addMembers = false;
        
        if (i_yapClassByClass.get(a_class) == null) {
            addYapClass(yapClass);
            addMembers = true;
        }
        
        int id = yapClass.getID();
        if(id == 0){
            yapClass.write(stream().getSystemTransaction());
            id = yapClass.getID();
        }
        
        if(i_yapClassByID.get(id) == null){
            i_yapClassByID.put(id, yapClass);
            addMembers = true;
        }
        
        if(addMembers || yapClass.i_fields == null){
			_classInits.process(yapClass);
        }
        
        i_creating.remove(a_class);
        
        stream().setDirtyInSystemTransaction(this);
        
        return yapClass;
    }    
    
	YapClass getYapClass(int id) {
        return readYapClass((YapClass)i_yapClassByID.get(id), null);
    }
	
    public YapClass getYapClass(String a_name) {
        YapClass yapClass = (YapClass)i_yapClassByBytes.remove(getNameBytes(a_name));
        readYapClass(yapClass, null);
        if (yapClass == null) {
            YapClassCollectionIterator i = iterator();
            while (i.moveNext()) {
                yapClass = (YapClass)i.current();
                if (a_name.equals(yapClass.getName())) {
                    readYapClass(yapClass, null);
                    return yapClass;
                }
            }
            return null;
        }
        return yapClass;
    }
    
    public int getYapClassID(String name){
        YapClass yc = (YapClass)i_yapClassByBytes.get(getNameBytes(name));
        if(yc != null){
            return yc.getID();
        }
        return 0;
    }

	byte[] getNameBytes(String name) {		
		return asBytes(resolveAliasRuntimeName(name));
	}

	private String resolveAliasRuntimeName(String name) {
		return stream().configImpl().resolveAliasRuntimeName(name);
	}

    void initOnUp(Transaction systemTrans) {
        i_yapClassCreationDepth++;
        systemTrans.stream().showInternalClasses(true);
        Iterator4 i = i_classes.iterator();
        while (i.moveNext()) {
            ((YapClass)i.current()).initOnUp(systemTrans);
        }
        systemTrans.stream().showInternalClasses(false);
        i_yapClassCreationDepth--;
        initYapClassesOnUp();
    }

    void initTables(int a_size) {
        i_classes = new Collection4();
        i_yapClassByBytes = new Hashtable4(a_size);
        if (a_size < 16) {
            a_size = 16;
        }
        i_yapClassByClass = new Hashtable4(a_size);
        i_yapClassByID = new Hashtable4(a_size);
        i_creating = new Hashtable4(1);
    }
    
    private void initYapClassesOnUp() {
        if(i_yapClassCreationDepth == 0){
            YapClass yc = (YapClass)i_initYapClassesOnUp.next();
            while(yc != null){
                yc.initOnUp(_systemTransaction);
                yc = (YapClass)i_initYapClassesOnUp.next();
            }
        }
    }
    
    public YapClassCollectionIterator iterator(){
        return new YapClassCollectionIterator(this, new ArrayIterator4(i_classes.toArray()));
    } 

    private static class ClassIDIterator extends MappingIterator {

		public ClassIDIterator(Collection4 classes) {
			super(classes.iterator());
		}
    	
    	protected Object map(Object current) {
    		return new Integer(((YapClass)current).getID());
    	}
    }
    
    public Iterator4 ids(){
        return new ClassIDIterator(i_classes);
    } 

    public int ownLength() {
        return YapConst.OBJECT_LENGTH
            + YapConst.INT_LENGTH
            + (i_classes.size() * YapConst.ID_LENGTH);
    }

    void purge() {
        Iterator4 i = i_classes.iterator();
        while (i.moveNext()) {
            ((YapClass)i.current()).purge();
        }
    }

    public final void readThis(Transaction a_trans, Buffer a_reader) {
		int classCount = a_reader.readInt();

		initTables(classCount);

		YapStream stream = stream();
		int[] ids = new int[classCount];

		for (int i = 0; i < classCount; ++i) {
			ids[i] = a_reader.readInt();
		}
		StatefulBuffer[] yapWriters = stream.readWritersByIDs(a_trans, ids);

		for (int i = 0; i < classCount; ++i) {
			YapClass yapClass = new YapClass(stream, null);
			yapClass.setID(ids[i]);
			i_classes.add(yapClass);
			i_yapClassByID.put(ids[i], yapClass);
			byte[] name = yapClass.readName1(a_trans, yapWriters[i]);
			i_yapClassByBytes.put(name, yapClass);
		}

		applyReadAs();

	}

	Hashtable4 classByBytes(){
    	return i_yapClassByBytes;
    }
    
    private void applyReadAs(){
        final Hashtable4 readAs = stream().configImpl().readAs(); 
        readAs.forEachKey(new Visitor4() {
            public void visit(Object a_object) {
                String dbName = (String)a_object;
                byte[] dbbytes = getNameBytes(dbName);
                String useName = (String)readAs.get(dbName);
                byte[] useBytes = getNameBytes(useName);
                if(classByBytes().get(useBytes) == null){
                    YapClass yc = (YapClass)classByBytes().get(dbbytes);
                    if(yc != null){
                        yc.i_nameBytes = useBytes;
                        yc.setConfig(stream().configImpl().configClass(dbName));
                        classByBytes().put(dbbytes, null);
                        classByBytes().put(useBytes, yc);
                    }
                }
            }
        });
    }

    public YapClass readYapClass(YapClass yapClass, ReflectClass a_class) {
    	if(yapClass == null){
    		return null;
    	}
        if (! yapClass.stateUnread()) {
            return yapClass;
        }
        i_yapClassCreationDepth++;
        yapClass.createConfigAndConstructor(i_yapClassByBytes, stream(), a_class);
        ReflectClass claxx = yapClass.classReflector();
        if(claxx != null){
            i_yapClassByClass.put(claxx, yapClass);
            yapClass.readThis();
            yapClass.checkChanges();
            i_initYapClassesOnUp.add(yapClass);
        }
        i_yapClassCreationDepth--;
        initYapClassesOnUp();
        return yapClass;
    }

    public void refreshClasses() {
        YapClassCollection rereader = new YapClassCollection(_systemTransaction);
        rereader.i_id = i_id;
        rereader.read(stream().getSystemTransaction());
        Iterator4 i = rereader.i_classes.iterator();
        while (i.moveNext()) {
            YapClass yc = (YapClass)i.current();
            if (i_yapClassByID.get(yc.getID()) == null) {
                i_classes.add(yc);
                i_yapClassByID.put(yc.getID(), yc);
                if(yc.stateUnread()){
                    i_yapClassByBytes.put(yc.readName(_systemTransaction), yc);
                }else{
                    i_yapClassByClass.put(yc.classReflector(), yc);
                }
            }
        }
        i = i_classes.iterator();
        while (i.moveNext()) {
            YapClass yc = (YapClass)i.current();
            yc.refresh();
        }
    }

    void reReadYapClass(YapClass yapClass){
        if(yapClass != null){
            reReadYapClass(yapClass.i_ancestor);
            yapClass.readName(_systemTransaction);
            yapClass.forceRead();
            yapClass.setStateClean();
            yapClass.bitFalse(YapConst.CHECKED_CHANGES);
            yapClass.bitFalse(YapConst.READING);
            yapClass.bitFalse(YapConst.CONTINUE);
            yapClass.bitFalse(YapConst.DEAD);
            yapClass.checkChanges();
        }
    }
    
    public StoredClass[] storedClasses() {
    	ensureAllClassesRead();
        StoredClass[] sclasses = new StoredClass[i_classes.size()];
        i_classes.toArray(sclasses);
        return sclasses;
    }

    public void writeAllClasses(){
        StoredClass[] storedClasses = storedClasses();
        for (int i = 0; i < storedClasses.length; i++) {
            YapClass yc = (YapClass)storedClasses[i];
            yc.setStateDirty();
        }
        
        for (int i = 0; i < storedClasses.length; i++) {
            YapClass yc = (YapClass)storedClasses[i];
            yc.write(_systemTransaction);
        }
    }

    public void writeThis(Transaction trans, Buffer a_writer) {
        a_writer.writeInt(i_classes.size());
        Iterator4 i = i_classes.iterator();
        while (i.moveNext()) {
            a_writer.writeIDOf(trans, i.current());
        }
    }

	public String toString(){
        if(! Debug4.prettyToStrings){
            return super.toString();
        }
		String str = "Active:\n";
		Iterator4 i = i_classes.iterator();
		while(i.moveNext()){
			YapClass yc = (YapClass)i.current();
			str += yc.getID() + " " + yc + "\n";
		}
		return str;
	}

    YapStream stream() {
        return _systemTransaction.stream();
    }
    
    public void setID(int a_id) {
    	if (stream().isClient()) {
    		super.setID(a_id);
    		return;
    	}
    	
        if(i_id == 0) {        	
			systemData().classCollectionID(a_id);
        }
        super.setID(a_id);
    }

	private SystemData systemData() {
		return _systemTransaction.i_file.systemData();
	}

}
